physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3.F90

```diff
@@ -0,0 +1,392 @@
+          errflg = 1
+          return
+       end if
+
+       if (do_shoc) then
+           write(errmsg,'(*(a))') 'SHOC is not currently compatible with GFDL MP v3'
+           errflg = 1
+           return
+       endif
+
+       call gfdl_cloud_microphys_v3_mod_init(me, master, nlunit, input_nml_file, logunit, fn_nml, hydrostatic, errmsg, errflg)
+
+       is_initialized = .true.
+
+   end subroutine gfdl_cloud_microphys_v3_init
+
+
+! =======================================================================
+!>\brief The subroutine 'gfdl_cloud_microphys_v3_finalize' terminates the GFDL
+!! cloud microphysics.
+!!
+!! \section arg_table_gfdl_cloud_microphys_v3_finalize  Argument Table
+!! \htmlinclude gfdl_cloud_microphys_v3_finalize.html
+!!
+   subroutine gfdl_cloud_microphys_v3_finalize(errmsg, errflg)
+
+       implicit none
+
+       character(len=*), intent(out) :: errmsg
+       integer,          intent(out) :: errflg
+
+   ! Initialize CCPP error handling variables
+       errmsg = ''
+       errflg = 0
+
+       if (.not.is_initialized) return
+
+       call gfdl_cloud_microphys_v3_mod_end()
+
+       is_initialized = .false.
+
+   end subroutine gfdl_cloud_microphys_v3_finalize
+
+!>\defgroup gfdlmp  GFDL Cloud Microphysics Module
+!! This is cloud microphysics package for GFDL global cloud resolving model.
+!! The algorithms are originally derived from Lin et al. (1983) \cite lin_et_al_1983.
+!! Most of the key elements have been simplified/improved. This code at this stage
+!! bears little to no similarity to the original Lin MP.
+!! Therefore, it is best to be called GFDL microphysics (GFDL MP) .
+!!
+!>\brief The module contains the GFDL cloud
+!! microphysics (Chen and Lin (2013) \cite chen_and_lin_2013 ).
+!> The module is paired with \ref fast_sat_adj, which performs the "fast"
+!! processes.
+!!
+!>\brief The subroutine executes the full GFDL cloud microphysics.
+!! \section arg_table_gfdl_cloud_microphys_v3_run Argument Table
+!! \htmlinclude gfdl_cloud_microphys_v3_run.html
+!!
+   subroutine gfdl_cloud_microphys_v3_run(                                         &
+      imp_physics, imp_physics_gfdl, fast_mp_consv,                                &
+      levs, im, rainmin, con_g, con_fvirt, con_rd, con_eps, garea, slmsk, snowd,   &
+      gq0, gq0_ntcw, gq0_ntrw, gq0_ntiw, gq0_ntsw, gq0_ntgl, gq0_ntclamt, aerfld,  &
+      gt0, gu0, gv0, vvl, prsl, phii, del,                                         &
+      rain0, ice0, snow0, graupel0, prcp0, sr, oro,                                &
+      dtp, hydrostatic, lradar, refl_10cm,                                         &
+      reset, effr_in, rew, rei, rer, res, reg,                                     &
+      cplchm, pfi_lsan, pfl_lsan, errmsg, errflg)
+
+      use machine, only: kind_phys, kind_dyn, kind_dbl_prec
+
+      implicit none
+
+      ! DH* TODO: CLEANUP, all of these should be coming in through the argument list
```

Comment:
Yes, and consider the comment RE: temperature_at_zero_celsius

Reply:
done.